### PR TITLE
Implement RFC 2580: Pointer metadata & VTable

### PR DESCRIPTION
**Pull request by SimonSapin**
RFC: https://github.com/rust-lang/rfcs/pull/2580

~~Before merging this PR:~~

* [x] Wait for the end of the RFC’s [FCP to merge](https://github.com/rust-lang/rfcs/pull/2580#issuecomment-759145278).
* [x] Open a tracking issue: https://github.com/rust-lang/rust/issues/81513
* [x] Update `#[unstable]` attributes in the PR with the tracking issue number

----

This PR extends the language with a new lang item for the `Pointee` trait which is special-cased in trait resolution to implement it for all types. Even in generic contexts, parameters can be assumed to implement it without a corresponding bound.

For this I mostly imitated what the compiler was already doing for the `DiscriminantKind` trait. I’m very unfamiliar with compiler internals, so careful review is appreciated.

This PR also extends the standard library with new unstable APIs in `core::ptr` and `std::ptr`:

```rust
pub trait Pointee {
    /// One of `()`, `usize`, or `DynMetadata<dyn SomeTrait>`
    type Metadata: Copy + Send + Sync + Ord + Hash + Unpin;
}

pub trait Thin = Pointee<Metadata = ()>;

pub const fn metadata<T: ?Sized>(ptr: *const T) -> <T as Pointee>::Metadata {}

pub const fn from_raw_parts<T: ?Sized>(*const (), <T as Pointee>::Metadata) -> *const T {}
pub const fn from_raw_parts_mut<T: ?Sized>(*mut (),<T as Pointee>::Metadata) -> *mut T {}

impl<T: ?Sized> NonNull<T> {
    pub const fn from_raw_parts(NonNull<()>, <T as Pointee>::Metadata) -> NonNull<T> {}

    /// Convenience for `(ptr.cast(), metadata(ptr))`
    pub const fn to_raw_parts(self) -> (NonNull<()>, <T as Pointee>::Metadata) {}
}

impl<T: ?Sized> *const T {
    pub const fn to_raw_parts(self) -> (*const (), <T as Pointee>::Metadata) {}
}

impl<T: ?Sized> *mut T {
    pub const fn to_raw_parts(self) -> (*mut (), <T as Pointee>::Metadata) {}
}

/// `<dyn SomeTrait as Pointee>::Metadata == DynMetadata<dyn SomeTrait>`
pub struct DynMetadata<Dyn: ?Sized> {
    // Private pointer to vtable
}

impl<Dyn: ?Sized> DynMetadata<Dyn> {
    pub fn size_of(self) -> usize {}
    pub fn align_of(self) -> usize {}
    pub fn layout(self) -> crate::alloc::Layout {}
}

unsafe impl<Dyn: ?Sized> Send for DynMetadata<Dyn> {}
unsafe impl<Dyn: ?Sized> Sync for DynMetadata<Dyn> {}
impl<Dyn: ?Sized> Debug for DynMetadata<Dyn> {}
impl<Dyn: ?Sized> Unpin for DynMetadata<Dyn> {}
impl<Dyn: ?Sized> Copy for DynMetadata<Dyn> {}
impl<Dyn: ?Sized> Clone for DynMetadata<Dyn> {}
impl<Dyn: ?Sized> Eq for DynMetadata<Dyn> {}
impl<Dyn: ?Sized> PartialEq for DynMetadata<Dyn> {}
impl<Dyn: ?Sized> Ord for DynMetadata<Dyn> {}
impl<Dyn: ?Sized> PartialOrd for DynMetadata<Dyn> {}
impl<Dyn: ?Sized> Hash for DynMetadata<Dyn> {}
```

API differences from the RFC, in areas noted as unresolved questions in the RFC:

* Module-level functions instead of associated `from_raw_parts` functions on `*const T` and `*mut T`, following the precedent of `null`, `slice_from_raw_parts`, etc.
* Added `to_raw_parts`

### CODE REVIEW

**Comment by SimonSapin** [comment]
What’s the correct syntax for intra-doc links to a method of a primitive type? `./x doc library/std --stage 1` gives:

```rust
error: unresolved link to `to_raw_parts`
  --> library/core/src/ptr/metadata.rs:51:6
   |
51 | /// [`to_raw_parts`]: <*const _>::to_raw_parts
   |      ^^^^^^^^^^^^^^ no item named `to_raw_parts` in scope

**Comment by jyn514** [comment]
> What’s the correct syntax for intra-doc links to a method of a primitive type? `./x doc library/std --stage 1` gives:
> 
> ```rust
> error: unresolved link to `to_raw_parts`
>   --> library/core/src/ptr/metadata.rs:51:6
>    |
> 51 | /// [`to_raw_parts`]: <*const _>::to_raw_parts
>    |      ^^^^^^^^^^^^^^ no item named `to_raw_parts` in scope
> ```

This was just implemented a few weeks ago and isn't available on beta: https://github.com/rust-lang/rust/pull/80181

**Comment by SimonSapin** [comment]
Thanks! Tests in that PR show that the answer (to the question I should have asked about the syntax for raw pointers specifically)  is `*const::foo`, not `<*const T>::foo`.

> isn't available on beta

That’s fine, the item with that doc-comment has `#[cfg(not(bootstrap))]` anyway since it depends on the new lang item.

**Comment by jyn514** [comment]
> Thanks! Tests in that PR show that the answer (to the question I should have asked about the syntax for raw pointers specifically) is `*const::foo`, not `<*const T>::foo`.

Note that pointers are about to be put behind a feature gate because there's no way to distinguish `*const` from `*mut` currently: https://github.com/rust-lang/rust/pull/81015

**Comment by SimonSapin** [comment]
Indeed, I’ve added `#![cfg_attr(not(bootstrap), feature(intra_doc_pointers))]`.

By the way, I’ve had to add it in *both* libcore and libstd. Presumably `pub use core::ptr;` causes those docs to be rendered again, and rustdoc looks at feature gates of "the current crate" rather than the crate that defines the item.

----

The `ui/panic-handler/weak-lang-item.rs` failure seems unrelated to the PR?

**Comment by jyn514** [comment]
> The ui/panic-handler/weak-lang-item.rs failure seems unrelated to the PR?

Yeah that one seems flaky, it failed in https://github.com/rust-lang/rust/pull/80965 too.

**Comment by KodrAus** [comment]
I've created a tracking issue: #81513

**Comment by SimonSapin** [comment]
The RFC is merged, tracking issue numbers are in, and CI is green. This is only waiting for review now.

File: compiler/rustc_middle/src/ty/sty.rs [hunk 1]
```diff
@@ -2110,6 +2116,54 @@ impl<'tcx> TyS<'tcx> {
         }
     }
 
+    /// Returns the type of metadata for (potentially fat) pointers to this type.
+    pub fn ptr_metadata_ty(&'tcx self, tcx: TyCtxt<'tcx>) -> Ty<'tcx> {
+        // FIXME: should this normalize?
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/sty.rs:2121]: we'll find out when things hit the `bug!` below

File: library/core/src/ptr/const_ptr.rs [hunk 2]
```diff
@@ -48,6 +48,17 @@ impl<T: ?Sized> *const T {
...
 
+    /// Decompose a (possibly wide) pointer into is address and metadata components.
+    ///
+    /// The pointer can be later reconstructed with [`from_raw_parts`].
+    #[cfg(not(bootstrap))]
+    #[unstable(feature = "ptr_metadata", issue = "81513")]
+    #[rustc_const_unstable(feature = "ptr_metadata", issue = "81513")]
+    #[inline]
+    pub const fn to_raw_parts(self) -> (*const (), <T as super::Pointee>::Metadata) {
+        (self.cast(), super::metadata(self))
```
Comment (oli-obk) [inline library/core/src/ptr/const_ptr.rs:59]:
```suggestion
        (self.cast(), metadata(self))
```

**Review by oli-obk — APPROVED** [review]
Apologies for taking so long. No more excuses today :laughing: 

r=me with the module path nit

**Comment by SimonSapin** [comment]
Thanks! Applied the suggestion and rebased. Conflicts were all automatically resolved by kdiff3.

**Comment by SimonSapin** [comment]
These "id is not unique" linkchecker failures seem unrelated?

File: library/core/src/ptr/metadata.rs [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+/// Raw pointers can be decomposed into the data address and metadata components
+/// with their [`to_raw_parts`] method.
+///
+/// Alternatively, metadata alone can be extracted with the [`metadata`] function.
+/// A reference can be passed to [`metadata`] and implicitly coerced.
+///
+/// A (possibly-wide) pointer can be put back together from its address and metadata
+/// with [`from_raw_parts`] or [`from_raw_parts_mut`].
+///
+/// [`to_raw_parts`]: *const::to_raw_parts
```
Comment (oli-obk) [inline library/core/src/ptr/metadata.rs:51]: I don't know this syntax, maybe rustdoc dies on some of this syntax somewhere (wondering due to the "Self" being mentioned)

**Review by oli-obk — COMMENTED** [review]
Not sure what could cause this kind of rustdoc error or how to debug it.

File: library/core/src/ptr/metadata.rs (hunk 3 above)
Comment (SimonSapin) [reply library/core/src/ptr/metadata.rs:51]: As discussed upthread this syntax is new: https://github.com/rust-lang/rust/pull/80181/files#diff-db4200f9416b7faa2360a2bf5faf63ca3bb81cb3a3819efd53fde86bd6fa4ebe

This link does work after `./x.py doc library/std --open --stage 2` on my machine, and it’s in `std/ptr/trait.Pointee.html` not in `std/keyword.Self.html`.

**Comment by jyn514** [comment]
@bors r-

This is unlikely to be spurious, there's a rustdoc bug here somewhere. @SimonSapin linking to pointers is still pretty buggy, I would avoid it here for now, use `../../pointer/method.whatever.html` for now. If you can post the file `doc/std/keyword.Self.html` that would be helpful for debugging the issue. If you could make an MCVE that would be even better.

**Comment by SimonSapin** [comment]
> If you could make an MCVE that would be even better.

That would be ideal of course, but I can not reproduce at all. I’ve rebased again just in case this is related to something merged into `master` very recently, and `./x.py test src/tools/linkchecker --stage 2` passes for me:

```
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.14s
	finished in 6.496 seconds
Build completed successfully in 0:00:35
```

I assume the line with lower-cased `finished` is from the linkchecker tool running.

By the way, stages appear to be off by one between builds and docs. Stage 1 is the first where `./x.py build` builds a standard library with `#[cfg(not(bootstrap))]`, but stage 2 is the first where `./x.py doc` documents a standard library built with `#[cfg(not(bootstrap))]`.

Here is `keyword.Self.html ` from my local build anyway: https://gist.github.com/SimonSapin/7d4763a08063ca7f2951e307618b6767. HTML IDs `search` and `rustdoc-vars` do appear to be unique in it, based on eyeballing with `grep`.

**Comment by ehuss** [comment]
The error is caused by the lack of case-sensitivity on the filesystem.  `std\keyword.Self.html` and `std\keyword.self.html` end up stomping on one another. Presumably you just had some bad luck where rustdoc was writing to both simultaneously, and it ended up corrupted, maybe?  I'm not too familiar with how rustdoc works.  Pretty sure it is unrelated to this PR.



> By the way, stages appear to be off by one between builds and docs.

Can you explain more what you saw?  If I run `x.py doc --stage=1 library/std`, I see the `not(bootstrap)` docs in `build/*/doc/std/...`.

**Comment by jyn514** [comment]
> The error is caused by the lack of case-sensitivity on the filesystem. std\keyword.Self.html and std\keyword.self.html end up stomping on one another. Presumably you just had some bad luck where rustdoc was writing to both simultaneously, and it ended up corrupted, maybe? I'm not too familiar with how rustdoc works. Pretty sure it is unrelated to this PR.

Ah ok, this is https://github.com/rust-lang/rust/issues/25879 then.

@bors r=oli-obk

**Comment by SimonSapin** [comment]
> Can you explain more what you saw? If I run `x.py doc --stage=1 library/std`, I see the `not(bootstrap)` docs in `build/*/doc/std/...`.

After running `./x.py doc library/std --open` on the branch of this PR I did not see in the browser any of the APIs introduced by this branch. Then same with `./x.py doc library/std --open --stage 1`. Only with `./x.py doc library/std --open --stage 2` did they show up. This is my config.toml:

```toml
profile = "compiler"
changelog-seen = 2

[build]
low-priority = true
```

**Comment by ehuss** [comment]
The default for `x.py doc` without a stage is 0 (not a default I would recommend).  On latest master, I just tried `./x.py doc library/std --open` and saw the API is missing and then `x.py doc library/std --open --stage 1` , and I see your new changes, so I'm not sure what happened.  Maybe there was a stale browser cache from running the previous command? 

I'm trying to think if there's some way a fingerprint error could cause it to fail to rebuild, but I can't think of a way to trigger that.  When developing rustc, the fingerprint Cargo uses for a stage1 compiler is not very good (because it does not contain the SHA hash, it can't tell when the compiler changes), but in this scenario I don't think that matters.

**Comment by SimonSapin** [comment]
> and then `x.py doc library/std --open --stage 1` , and I see your new changes

Ok, it’s possible I made some mistake on my end then. Thanks for trying to reproduce.

**Comment by yshui** [comment]
@SimonSapin  Why isn't `DynMetadata` `repr(C)`? What is the intended way of passing the metadata through FFI and back?

**Comment by bjorn3** [comment]
Responded on the tracking issuen

**Pull request merged**